presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapColumnWriter.java

```diff
@@ -0,0 +1,379 @@
+    private final int stringStatisticsLimitInBytes;
+
+    private final LongOutputStream lengthStream;
+    private final PresentOutputStream presentStream;
+    private List<FlatMapKeyNodeWriter> keyNodes;
+    /* Block Writer for Numeric or String Keyed Types */
+    private final FlatMapBlockWriter blockWriter;
+    private StatisticsBuilder statisticsBuilder;
+    private final Supplier<StatisticsBuilder> statisticsBuilderSupplier;
+
+    private final List<ColumnStatistics> rowGroupColumnStatistics = new ArrayList<>();
+    private final List<ColumnStatistics> keyRowGroupColumnStatistics = new ArrayList<>();
+    private long columnStatisticsRetainedSizeInBytes;
+
+    private int nonNullValueCount;
+    private boolean closed;
+
+    private StringStatisticsBuilder newStringStatisticsBuilder()
+    {
+        return new StringStatisticsBuilder(stringStatisticsLimitInBytes);
+    }
+
+    public FlatMapColumnWriter(
+            int column,
+            int sequence,
+            ColumnWriterOptions columnWriterOptions,
+            Optional<DwrfDataEncryptor> dwrfEncryptor,
+            OrcEncoding orcEncoding,
+            MetadataWriter metadataWriter,
+            Type type,
+            List<OrcType> orcTypes,
+            DwrfSequenceColumnWriterFactory valueWriterfactory)
+    {
+        checkArgument(column >= 0, "column is negative");
+        checkArgument(sequence >= 0, "sequence is negative");
+        checkArgument(orcEncoding == DWRF, "Flat Maps are only supported for DWRF Encoding");
+
+        this.column = column;
+        this.sequence = sequence;
+        this.keyColumnId = orcTypes.get(column).getFieldTypeIndex(0);
+        this.valueColumnId = orcTypes.get(column).getFieldTypeIndex(1);
+
+        this.type = (MapType) type;
+        this.keyType = type.getTypeParameters().get(0);
+        this.valueType = type.getTypeParameters().get(1);
+
+        this.keyOrcType = orcTypes.get(keyColumnId);
+        this.valueOrcType = orcTypes.get(valueColumnId);
+        this.compressed = columnWriterOptions.getCompressionKind() != NONE;
+
+        this.columnEncoding = new ColumnEncoding(DWRF_MAP_FLAT, 0);
+        this.keyColumnEncoding = new ColumnEncoding(DIRECT, 0);
+
+        if (this.valueOrcType.getOrcTypeKind() == MAP) {
+            this.valueColumnEncoding = new ColumnEncoding(DWRF_MAP_FLAT, 0);
+        }
+        else {
+            this.valueColumnEncoding = new ColumnEncoding(DIRECT, 0);
+        }
+
+        this.dwrfEncryptor = dwrfEncryptor;
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, columnWriterOptions, dwrfEncryptor);
+        this.inputMetadataWriter = metadataWriter;
+
+        this.valueWriterFactory = valueWriterfactory;
+        this.stringStatisticsLimitInBytes = toIntExact(columnWriterOptions.getStringStatisticsLimit().toBytes());
+
+        this.lengthStream = createLengthOutputStream(columnWriterOptions, dwrfEncryptor, orcEncoding);
```

Comment:
does flat map have length stream ? Looks like we use it to checkpoint the index, but never flush the contents to disk. This looks incorrect to me.